webui/react/src/pages/ExperimentDetails/ExperimentVisualization/HpScatterPlots.tsx

```diff
@@ -0,0 +1,239 @@
+
+  const resetData = useCallback(() => {
+    setChartData(undefined);
+    setHasLoaded(false);
+  }, []);
+
+  const handleBatchChange = useCallback((batch: SelectValue) => {
+    if (!onBatchChange) return;
+    resetData();
+    onBatchChange(batch as number);
+  }, [ onBatchChange, resetData ]);
+
+  const handleHParamChange = useCallback((hps: SelectValue) => {
+    if (!onHParamChange) return;
+    if (Array.isArray(hps)) {
+      onHParamChange(hps.length === 0 ? undefined : hps as string[]);
+    }
+  }, [ onHParamChange ]);
+
+  const handleMetricChange = useCallback((metric: MetricName) => {
+    if (!onMetricChange) return;
+    resetData();
+    onMetricChange(metric);
+  }, [ onMetricChange, resetData ]);
+
+  useEffect(() => {
+    const canceler = new AbortController();
+
+    const trialIds: number[] = [];
+    const hpTrialMap: Record<string, Record<number, { hp: number, metric: number }>> = {};
+
+    consumeStream<V1TrialsSnapshotResponse>(
+      detApi.StreamingInternal.determinedTrialsSnapshot(
+        experiment.id,
+        selectedBatch,
+        selectedMetric.name,
+        metricTypeParamMap[selectedMetric.type],
+        undefined,
+        { signal: canceler.signal },
+      ),
+      event => {
+        if (!event || !event.trials || !Array.isArray(event.trials)) return;
+
+        const hpMetricMap: Record<string, number[]> = {};
+        const hpValueMap: Record<string, number[]> = {};
+        const hpLogScaleMap: Record<string, boolean> = {};
+
+        event.trials.forEach(trial => {
+          const trialId = trial.trialId;
+          trialIds.push(trialId);
+
+          hParams.forEach(hParam => {
+            if (!isNumber(trial.hparams[hParam])) return;
+
+            hpTrialMap[hParam] = hpTrialMap[hParam] || {};
+            hpTrialMap[hParam][trialId] = hpTrialMap[hParam][trialId] || {};
+            hpTrialMap[hParam][trialId] = {
+              hp: trial.hparams[hParam],
+              metric: trial.metric,
+            };
+          });
+        });
+
+        hParams.forEach(hParam => {
+          const hp = (experiment.config.hyperparameters || {})[hParam];
+          if (hp.type === ExperimentHyperParamType.Log) hpLogScaleMap[hParam] = true;
+
+          hpMetricMap[hParam] = [];
+          hpValueMap[hParam] = [];
+          trialIds.forEach(trialId => {
+            hpMetricMap[hParam].push(hpTrialMap[hParam][trialId].metric);
+            hpValueMap[hParam].push(hpTrialMap[hParam][trialId].hp);
+          });
+        });
+
+        setChartData({
+          hpLogScales: hpLogScaleMap,
+          hpValues: hpValueMap,
+          metricValues: hpMetricMap,
+          trialIds,
+        });
+        setHasLoaded(true);
+      },
+    ).catch(e => setPageError(e));
+
+    return () => canceler.abort();
+  }, [ experiment, hParams, selectedBatch, selectedMetric ]);
+
+  if (pageError) {
+    return <Message title={pageError.message} />;
+  } else if (hasLoaded && !chartData) {
```

Comment:
no need for that `else` (out of code readability)?

Reply:
yeah exactly, just a little bit more succinct. More so when we have more than two paths.